tests/lib/lazy_orm.py

```diff
@@ -0,0 +1,90 @@
+ORM provides _almost_ enough abstraction to let us do this, but not quite. The sticking
+point is that we need to be able to customise the type of some columns on a per-database
+basis. For instance, when inserting into MSSQL we need `MSSQLDate()` and when inserting
+into Spark we need `SparkDate()`. SQLAlchemy itself provides no mechanism that I can see
+for doing this.
+
+The solution I have implemented below is a "lazy ORM" class. These looks pretty much
+like regular ORM classes when you define them, and their instances act just enough like
+ORM objects that you can build up a set a fixtures with them; but it's not until you
+"hydrate" these objects that they become actual ORM instances. This gives us the
+opportunity to customise the hydration process for each database and produce ORM classes
+with the appropriate column types.
+
+This is, obviously, a terrible thing. But it's at least terrible in a sort of generic
+Python way rather than reaching deep into the internals of SQLAlchemy. So I felt on
+balance that it was the least-worst of the many terrible solutions I tried.
+
+It also has the advantage of being reasonably self-contained and easy to back out: the
+lazy ORM classes are just a find-and-replace away from being regular ORM classes; and
+the hydration class is used in only a single place.
+"""
+import functools
+
+import sqlalchemy
+
+
+class LazyBase:
+    def __init_subclass__(cls, **kwargs):
+        # Keep a record of defined columns
+        cls._columns = {}
+        for key, value in vars(cls).items():
+            if isinstance(value, LazyColumn):
+                cls._columns[key] = value
+                # If we don't do this, instances of the class will have the column
+                # definition as their default column value, which is not what we want
+                setattr(cls, key, None)
+            elif isinstance(value, sqlalchemy.Column):
+                assert False, f"'{key}' must be a LazyColumn not a sqlalchemy.Column"
+
+    def __init__(self, **kwargs):
+        # This is almost verbatim what the default ORM constructor does
+        for key, value in kwargs.items():
+            if key not in self._columns:
+                raise TypeError(f"{key} is not a valid keyword argument")
+            setattr(self, key, value)
+
+
+class LazyColumn:
+    """
+    This does nothing apart from record the arguments to the column constructor so we
+    can use them to create an actual `sqlalchemy.Column` later (possibly after
+    customising the column definition to suit the target database).
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+
+
+class HydrateLazyObjects:
+    def __init__(self):
+        self.orm_base_class = sqlalchemy.orm.declarative_base()
+        # Use a cache to make sure we get the same corresponding ORM class for each
+        # LazyBase subclass instance we see
+        self.get_orm_class = functools.cache(self.make_orm_class)
```

Comment:
Why isn't this a decorator on `make_orm_class()`?

Reply:
Ah, I should have explained this better in the comment. We have to decorate the method on the instance rather than the class otherwise we get a single cache shared between all instances and the point of the code is to create _different_ classes per-instance.

Reply:
Huh. So `functools.cache` holds a single cache for all instances of the class, when it's used as a decorator? That seems surprising.

Reply:
So I just tried this and was very surprised because it works exactly as you suggested (each instance gets its own cache) but I couldn't for the life of me figure out how because we only decorate the function once at definition time and `functools.cache` has no special handling for instance methods.

Then I eventually realised that they _do_ all share the same cache, it's just that the key is the full list of arguments which includes `self`; so this makes the single shared cache effectively partitioned by instance. That's probably obvious in retrospect, but it certainly confused me for a while.

I'll update the code and add a note.

Reply:
I too read the `functools` code and I missed the use of `self` in the key.